storage/src/nodestore/persist.rs

```diff
@@ -74,6 +80,117 @@ impl<T, S: WritableStorage> NodeStore<T, S> {
+    /// Creates a new iterator over unpersisted nodes in depth-first order.
+    fn new(store: &'a N) -> Self {
+        let root = store.root_as_maybe_persisted_node();
+
+        // we must have an unpersisted root node to use this iterator
+        // It's hard to tell at compile time if this is the case, so we assert it here
+        // TODO: can we use another trait or generic to enforce this?
+        debug_assert!(root.as_ref().is_none_or(|r| r.unpersisted().is_some()));
+        let (child_iter_stack, stack) = if let Some(root) = root {
+            if let Some(branch) = root
+                .as_shared_node(store)
+                .expect("in memory, so no io")
+                .as_branch()
+            {
+                // Create an iterator over unpersisted children
+                let unpersisted_children: Vec<MaybePersistedNode> = branch
+                    .children
+                    .iter()
+                    .filter_map(|child_opt| {
+                        child_opt
+                            .as_ref()
+                            .and_then(|child| child.unpersisted().cloned())
+                    })
+                    .collect();
+
+                (
+                    vec![Box::new(unpersisted_children.into_iter())
+                        as Box<dyn Iterator<Item = MaybePersistedNode> + 'a>],
+                    vec![root],
+                )
+            } else {
+                // root is a leaf
+                (vec![], vec![root])
+            }
+        } else {
+            (vec![], vec![])
+        };
+
+        Self {
+            store,
+            stack,
+            child_iter_stack,
+        }
+    }
+}
+
+#[cfg(test)]
+impl<N: NodeReader + RootReader> Iterator for UnPersistedNodeIterator<'_, N> {
+    type Item = MaybePersistedNode;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // Try to get the next child from the current child iterator
+        while let Some(current_iter) = self.child_iter_stack.last_mut() {
+            if let Some(next_child) = current_iter.next() {
+                let shared_node = next_child
+                    .as_shared_node(self.store)
+                    .expect("in memory, so IO is impossible");
+
+                if shared_node.is_leaf() {
+                    return Some(next_child);
+                }
+
+                // It's a branch, so we need to get its children
+                if let Some(branch) = shared_node.as_branch() {
+                    // Create an iterator over unpersisted children
+                    let unpersisted_children: Vec<MaybePersistedNode> = branch
+                        .children
+                        .iter()
+                        .filter_map(|child_opt| {
+                            child_opt
+                                .as_ref()
+                                .and_then(|child| child.unpersisted().cloned())
+                        })
```

Comment:
may be add a `rev()` to make sure that childrens are iterated in increasing order.

Reply:
The child order doesn't actually matter. In fact ,this could be done in parallel in a future PR.